libnetwork/datastore/datastore.go

```diff
@@ -261,6 +261,29 @@ func (ds *Store) Map(key string, kvObject KVObject) (map[string]KVObject, error)
 	return results, nil
 }
 
+// DeleteObjectIdempotent deletes a kvObject from the on-disk DB and the
+// in-memory cache. Unlike DeleteObjectAtomic, it doesn't check the optimistic
+// lock of the passed kvObject.
+func (ds *Store) DeleteObjectIdempotent(kvObject KVObject) error {
+	ds.mu.Lock()
+	defer ds.mu.Unlock()
+
+	if kvObject == nil {
+		return types.InvalidParameterErrorf("invalid KV Object : nil")
+	}
+
+	if !kvObject.Skip() {
+		if err := ds.store.DeleteIdempotent(Key(kvObject.Key()...)); err != nil {
+			return err
+		}
+	}
+
+	// cleanup the cache only if AtomicDelete went through successfully
+	// If persistent store is skipped, sequencing needs to
+	// happen in cache.
+	return ds.cache.del(kvObject, false)
+}
+
 // DeleteObjectAtomic performs atomic delete on a record.
 func (ds *Store) DeleteObjectAtomic(kvObject KVObject) error {
```

Comment:
Do we actually still _need_ / _use_ DeleteObjectAtomic ? It looks like the remaining use is `deleteFromStore`, which also does a `retry`; https://github.com/moby/moby/blob/c42ae61e627b2e6514ea4afbb4cd8f02ce09ed90/libnetwork/store.go#L170-L183


- ❓ Does that mean we're always "force-deleting"?
- ❓ are there cases where "force-deleting" could be problematic? (i.e. we delete because nothing is attached anymore to a network, but there's a race, and now we ignore that something is attached?

Reply:
>❓ are there cases where "force-deleting" could be problematic? (i.e. we delete because nothing is attached anymore to a network, but there's a race, and now we ignore that something is attached?

It's even worse than that. This retry loop refreshes the object it tries to delete if it's been updated by another thread in the mean time. When `deleteFromStore` returns, the callsite gets the "refreshed" object and might rely on that side-effect to do whatever it has to do.

To make sure we won't introduce new bugs, we need to first ensure that this side-effect isn't needed for the affected callsites to do their job, or to make sure this side effect won't ever exists (through better locking).

In that regard, the callsites affected by this PR are pretty easy to verify. But `deleteFromStore` callsites are harder to review and left untouched purposefully.